File: src/main/java/fi/metatavu/metaform/server/exporttheme/ExportThemeFreemarkerTemplateLoader.java
package fi.metatavu.metaform.server.exporttheme;

import java.io.IOException;
import java.io.Reader;
import java.io.StringReader;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;

import org.slf4j.Logger;

import fi.metatavu.metaform.server.persistence.model.ExportThemeFile;
import freemarker.cache.TemplateLoader;

/**
 * Freemarker template loader for loading templates from database
 * 
 * @author Antti Leppä
 */
@ApplicationScoped
public class ExportThemeFreemarkerTemplateLoader implements TemplateLoader {
  
  @Inject
  private Logger logger;

  @Inject
  private ExportThemeController exportThemeController;
  
  @Override
  public Object findTemplateSource(String name) {
    return name;
  }

  @Override
  public long getLastModified(Object templateSource) {
    String path = (String) templateSource;
    ExportThemeFile exportThemeFile = findExportThemeFile(path);
    if (exportThemeFile != null) {
      return exportThemeFile.getModifiedAt().toInstant().toEpochMilli();
    }
    
    return 0;
  }

  @Override
  public Reader getReader(Object templateSource, String encoding) throws IOException {
    String path = (String) templateSource;
    ExportThemeFile exportThemeFile = findExportThemeFile(path);
    if (exportThemeFile != null) {
      return new StringReader(exportThemeFile.getContent());
    }
    
    logger.warn("Could not find export theme file {}", path);
    
    return new StringReader(String.format("!! export theme file %s not found !!", path));
  }
  
  private ExportThemeFile findExportThemeFile(String path) {
    
    
    return exportThemeController.findExportThemeFile(path);
  }

  @Override
  public void closeTemplateSource(Object templateSource) throws IOException {
    // Template loader is id, so no need to close
  }
  
}

File: src/main/kotlin/fi/metatavu/metaform/server/exportTheme/ExportThemeFreemarkerTemplateLoader.kt
package fi.metatavu.metaform.server.exportTheme

import fi.metatavu.metaform.server.controllers.ExportThemeController
import fi.metatavu.metaform.server.persistence.model.ExportThemeFile
import freemarker.cache.TemplateLoader
import org.slf4j.Logger
import java.io.IOException
import java.io.Reader
import java.io.StringReader
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject

/**
 * Freemarker template loader for loading templates from database
 *
 * @author Antti Leppä
 */
@ApplicationScoped
class ExportThemeFreemarkerTemplateLoader: TemplateLoader {
    @Inject
    lateinit var logger: Logger

    @Inject
    lateinit var exportThemeController: ExportThemeController

    override fun findTemplateSource(name: String?): Any? {
        return name
    }

    override fun getLastModified(templateSource: Any?): Long {
        val path = templateSource as String?
        val exportThemeFile = findExportThemeFile(path)
        return if (exportThemeFile != null) {
            exportThemeFile.modifiedAt!!.toInstant().toEpochMilli()
        } else 0
    }

    @Throws(IOException::class)
    override fun getReader(templateSource: Any?, encoding: String?): Reader? {
        val path = templateSource as String
        val exportThemeFile = findExportThemeFile(path)
        if (exportThemeFile != null) {
            return StringReader(exportThemeFile.content)
        }
        val baseReader = exportThemeController.findBaseThemeWithinJar(path)?.bufferedReader()
        if (baseReader != null) {
            return baseReader
        }
        logger.warn("Could not find export theme file {}", path)
        return StringReader(String.format("!! export theme file %s not found !!", path))
    }

    override fun closeTemplateSource(templateSource: Any?) {
        // Template loader is id, so no need to close
    }

    /**
     * Finds export theme file by path
     *
     * @param path path
     * @return found export theme file or null if not found
     */
    private fun findExportThemeFile(path: String?): ExportThemeFile? {
        return exportThemeController.findExportThemeFile(path = path)
    }
}